### PR TITLE
Add fourier noise function

### LINKED ISSUE #13316: Fourier series based random noise
## Reason
For phase field simulations noise that is added should have a frequency cutoff related to the interfacial width. This could be done by filtering existing noise, but a better approach is to generate the noise in fourier space to begin with. This noise will also be completely grid independent and can be engineered to have an arbitrary power spectrum.

## Design
Add a Moose Function object that generates a new fourier series for a random noise field with a 
frequency cut off at the beginning of a timestep. The whole series must be generated on each MPI rank, as k-space cannot be partitioned like the underlying real space. This will automatically lead to parallel reproducible (diffuse) noise fields.

## Impact
Added capability for noise based nucleation. (following Gránásy _et al._)

### PR DESCRIPTION
**Pull request by dschwen**
Closes #13316 

Tag @mangerij

### CODE REVIEW

**Comment by dschwen** [comment]
No!

**Comment by mangerij** [comment]
hey excited to test it out :)

File: modules/phase_field/src/functions/FourierNoise.C [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                 "If specifying the number of terms, supply a number greater than zero.");
+
+    // fill terms
+    for (auto & f : _series)
+    {
+      // get a vector with length <= 0.5
+      Real r2;
+      do
+      {
+        const Real x = MooseRandom::rand() - 0.5;
```
Comment (permcody) [inline modules/phase_field/src/functions/FourierNoise.C]: Why are you using the static instance here? I thought you wanted to exterminate it. I would suggest using a local instance so that the inclusion of this object doesn't change behaviors of other objects that still use the static version.

Comment (dschwen) [reply modules/phase_field/src/functions/FourierNoise.C]: That local instance is just so kludgy. The hash map is superfluous when you can just create multiple instances of MooseRandom...

Comment (permcody) [reply modules/phase_field/src/functions/FourierNoise.C]: OK - so the interface isn't perfect, I agree. Perhaps we need an overload to just use some default state in MooseRandom that isn't backed up or restored (e.g. it behaves just like the static instance except it's not global. Would that satisfy your needs? I'd like to see the use of the static generator be discouraged. We have code all over the framework and module that interact with the static instance. Each one of them feeding seeds and getting out random numbers. In some cases we reseed regularly, which is kind of dangerous to your desire for good random numbers. As you add or remove these objects from your sim, you impact behavior, which I think is possibly worse than using a slightly clunky interface.

TL;DR; Let's fix up the interface to MooseRandom and avoid the static usage unless you are diametrically opposed.

Comment (permcody) [reply modules/phase_field/src/functions/FourierNoise.C]: Also - there's this: https://github.com/idaholab/moose/issues/5567

Comment (dschwen) [reply modules/phase_field/src/functions/FourierNoise.C]: Can you have an overload with the same function signatures except one is `static` and the other one isn't?!

Comment (dschwen) [reply modules/phase_field/src/functions/FourierNoise.C]: No, directly prohibited by the standard:

https://stackoverflow.com/questions/5365689/c-overload-static-function-with-non-static-function

Comment (dschwen) [reply modules/phase_field/src/functions/FourierNoise.C]: Mys suggestion would be to remove static usage and keep the function signatures without the index parameter to refer to a separate state that is not stored in the map. It'll be a bit of a ripping-off-the-bandaid-pain to regold some tests, and fix some apps, but worth it IMO.

Comment (permcody) [reply modules/phase_field/src/functions/FourierNoise.C]: I agree with that. I'm not sure I want to completely remove the static usage though. Perhaps we rename the static methods and maybe even print a warning when they are used? The static methods are still better than using system methods, but only just barely.

Comment (dschwen) [reply modules/phase_field/src/functions/FourierNoise.C]: nonetheless, this is all orthogonal to this PR

**Review by permcody — APPROVED** [review]
